pycbc/waveform/waveform_modes.py

```diff
@@ -0,0 +1,491 @@
+    polarization.
+
+    Parameters
+    ----------
+    hlms : dict
+        Dictionary of ``(l, m)`` -> complex ``hlm``. The ``hlm`` may be a
+        complex number or array, or complex ``TimeSeries``. All modes in the
+        dictionary will be summed.
+    inclination : float
+        The inclination to use.
+    phi : float
+        The phase to use.
+
+    Returns
+    -------
+    complex float or array
+        The plus and cross polarization as a complex number. The real part
+        gives the plus, the negative imaginary part the cross.
+    """
+    out = None
+    for mode in hlms:
+        ell, m = mode
+        hlm = hlms[ell, m]
+        ylm = lal.SpinWeightedSphericalHarmonic(inclination, phi, -2, ell, m)
+        if out is None:
+            out = ylm * hlm
+        else:
+            out += ylm * hlm
+    return out
+
+
+def default_modes(approximant):
+    """Returns the default modes for the given approximant.
+    """
+    # FIXME: this should be replaced to a call to a lalsimulation function,
+    # whenever that's added
+    if approximant in ['IMRPhenomXPHM', 'IMRPhenomXHM']:
+        # according to arXiv:2004.06503
+        ma = [(2, 2), (2, 1), (3, 3), (3, 2), (4, 4)]
+        # add the -m modes
+        ma += [(ell, -m) for ell, m in ma]
+    elif approximant in ['IMRPhenomPv3HM', 'IMRPhenomHM']:
+        # according to arXiv:1911.06050
+        ma = [(2, 2), (2, 1), (3, 3), (3, 2), (4, 4), (4, 3)]
+        # add the -m modes
+        ma += [(ell, -m) for ell, m in ma]
+    elif approximant.startswith('NRSur7dq4'):
+        # according to arXiv:1905.09300
+        ma = [(ell, m) for ell in [2, 3, 4] for m in range(-ell, ell+1)]
+    else:
+        raise ValueError("I don't know what the default modes are for "
+                         "approximant {}, sorry!".format(approximant))
+    return ma
+
+
+def get_glm(ell, m, theta):
+    """The inclination part of the ylm."""
+    return lal.SpinWeightedSphericalHarmonic(theta, 0., -2, ell, m).real
+
+
+def get_nrsur_modes(**params):
+    """Generates NRSurrogate waveform mode-by-mode.
+
+    All waveform parameters should be provided as keyword arguments.
+    Recognized parameters are listed below. Unrecognized arguments are ignored.
+
+    Parameters
+    ----------
+    template: object
+        An object that has attached properties. This can be used to substitute
+        for keyword arguments. A common example would be a row in an xml table.
+    approximant : str
+        The approximant to generate. Must be one of the ``NRSur*`` models.
+    {delta_t}
+    {mass1}
+    {mass2}
+    {spin1x}
+    {spin1y}
+    {spin1z}
+    {spin2x}
+    {spin2y}
+    {spin2z}
+    {f_lower}
+    {f_ref}
+    {distance}
+    {mode_array}
+
+    Returns
+    -------
+    dict :
+        Dictionary of ``(l, m)`` -> ``(h_+, -h_x)`` ``TimeSeries``.
+    """
+    laldict = _check_lal_pars(params)
+    ret = lalsimulation.SimInspiralPrecessingNRSurModes(
```

Comment:
I don't like that we should have to define this for every waveform model. I'm reasonably sure that `lalsimulation` defines a generic `get_specific_modes` function (although probably not all waveforms are plugged into it). Did this never actually happen, or is it not usable for some reason?

Reply:
There is no such infrastructure at the moment, though there appear to be some remnants of past efforts. In particular there are partial functions for XPHM and Pv3HM which either outright don't work and are disable, or use an inconsistent convention which is not possible to combine with spherical harmonics afterwards.

Reply:
@spxiwh We've confirmed with Seb that we haven't missed something and have raised our support for a generic interface in the future, which would make it easier for us to pick up lalsimulation models. Of course, we'd like to be generic and allow for models outside lalsimulation, but it is indeed  a pain if they all use different conventions.

Reply:
Fair enough.